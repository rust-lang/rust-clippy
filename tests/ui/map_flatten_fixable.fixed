// run-rustfix

#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::let_underscore_drop)]
#![allow(clippy::missing_docs_in_private_items)]
#![allow(clippy::map_identity)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::unnecessary_wraps)]
#![feature(result_flattening)]

fn main() {
    // mapping to Option on Iterator
    fn option_id(x: i8) -> Option<i8> {
        Some(x)
    }
    let option_id_ref: fn(i8) -> Option<i8> = option_id;
    let option_id_closure = |x| Some(x);
    let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id).collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id_ref).collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id_closure).collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(|x| x.checked_add(1)).collect();

    // mapping to Iterator on Iterator
    let _: Vec<_> = vec![5_i8; 6].into_iter().flat_map(|x| 0..x).collect();

    // mapping to Option on Option
    let _: Option<_> = (Some(Some(1))).and_then(|x| x);

    // mapping to Result on Result
    let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
}
