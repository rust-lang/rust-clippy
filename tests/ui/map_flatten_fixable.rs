// run-rustfix

#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::let_underscore_drop)]
#![allow(clippy::missing_docs_in_private_items)]
#![allow(clippy::map_identity)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::unnecessary_wraps)]
#![feature(result_flattening)]

fn main() {
    // mapping to Option on Iterator
    fn option_id(x: i8) -> Option<i8> {
        Some(x)
    }
    let option_id_ref: fn(i8) -> Option<i8> = option_id;
    let option_id_closure = |x| Some(x);
    let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id).flatten().collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id_ref).flatten().collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id_closure).flatten().collect();
    let _: Vec<_> = vec![5_i8; 6].into_iter().map(|x| x.checked_add(1)).flatten().collect();

    // mapping to Iterator on Iterator
    let _: Vec<_> = vec![5_i8; 6].into_iter().map(|x| 0..x).flatten().collect();

    // mapping to Option on Option
    let _: Option<_> = (Some(Some(1))).map(|x| x).flatten();

    // mapping to Result on Result
    let _: Result<_, &str> = (Ok(Ok(1))).map(|x| x).flatten();
}
