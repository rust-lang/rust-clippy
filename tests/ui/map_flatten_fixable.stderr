error: called `map(..).flatten()` on `Iterator`
  --> $DIR/map_flatten_fixable.rs:18:47
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id).flatten().collect();
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::map-flatten` implied by `-D warnings`
help: try replacing `map` with `filter_map`, and remove the `.flatten()`
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id).collect();
   |                                               ~~~~~~~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Iterator`
  --> $DIR/map_flatten_fixable.rs:19:47
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id_ref).flatten().collect();
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `filter_map`, and remove the `.flatten()`
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id_ref).collect();
   |                                               ~~~~~~~~~~~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Iterator`
  --> $DIR/map_flatten_fixable.rs:20:47
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().map(option_id_closure).flatten().collect();
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `filter_map`, and remove the `.flatten()`
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(option_id_closure).collect();
   |                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Iterator`
  --> $DIR/map_flatten_fixable.rs:21:47
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().map(|x| x.checked_add(1)).flatten().collect();
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `filter_map`, and remove the `.flatten()`
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().filter_map(|x| x.checked_add(1)).collect();
   |                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Iterator`
  --> $DIR/map_flatten_fixable.rs:24:47
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().map(|x| 0..x).flatten().collect();
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `flat_map`, and remove the `.flatten()`
   |
LL |     let _: Vec<_> = vec![5_i8; 6].into_iter().flat_map(|x| 0..x).collect();
   |                                               ~~~~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Option`
  --> $DIR/map_flatten_fixable.rs:27:40
   |
LL |     let _: Option<_> = (Some(Some(1))).map(|x| x).flatten();
   |                                        ^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `and_then`, and remove the `.flatten()`
   |
LL |     let _: Option<_> = (Some(Some(1))).and_then(|x| x);
   |                                        ~~~~~~~~~~~~~~~

error: called `map(..).flatten()` on `Result`
  --> $DIR/map_flatten_fixable.rs:30:42
   |
LL |     let _: Result<_, &str> = (Ok(Ok(1))).map(|x| x).flatten();
   |                                          ^^^^^^^^^^^^^^^^^^^^
   |
help: try replacing `map` with `and_then`, and remove the `.flatten()`
   |
LL |     let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
   |                                          ~~~~~~~~~~~~~~~

error: aborting due to 7 previous errors

